sdk/typespec/typespec_client_core/examples/binary_data_request.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use tokio::fs;
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::fmt::format::FmtSpan;
+use typespec_client_core::{fs::FileStreamBuilder, http::RequestContent};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Log traces to stdout.
+    let _log = tracing_subscriber::fmt()
+        .with_span_events(FmtSpan::ENTER | FmtSpan::EXIT)
+        .with_max_level(LevelFilter::DEBUG)
+        .init();
+
+    // Asynchronously read the whole file into memory.
+    let body = RequestContent::from(fs::read(file!()).await?);
+    client::put_binary_data(body).await?;
+
+    // Asynchronously stream the file with the service client request.
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        let file = fs::File::open(file!()).await?;
+        let file = FileStreamBuilder::new(file)
+            // Simulate a slow, chunky request.
+            .buffer_size(256usize)
+            .build()
+            .await?;
+        client::put_binary_data(file.into()).await?;
+    }
+
+    Ok(())
+}
+
+mod client {
+    use futures::StreamExt;
+    use tracing::debug;
+    use typespec_client_core::{
+        http::{headers::Headers, Body, RequestContent, Response, StatusCode},
+        stream::BytesStream,
+    };
+
+    #[tracing::instrument(skip(body))]
+    pub async fn put_binary_data(
+        body: RequestContent<()>,
```

Comment:
No, body should be just `RequestContent<T>` since that has all the stuff to go `From` other types; besides, I'm leaning toward taking concrete parameter types. See the internal thread I CC'd you et. al. on.